google/cloud/spanner/internal/connection_impl.cc

```diff
@@ -389,19 +398,22 @@ RowStream ConnectionImpl::ReadImpl(
         backoff_policy_prototype_->clone());
     auto reader = PartialResultSetSource::Create(std::move(rpc));
     if (s->has_begin()) {
-      auto metadata = reader.ok() ? (*reader)->Metadata() : absl::nullopt;
-      if (metadata && metadata->has_transaction()) {
-        s->set_id(metadata->transaction().id());
+      if (reader.ok()) {
+        auto metadata = (*reader)->Metadata();
+        if (metadata && metadata->has_transaction()) {
+          s->set_id(metadata->transaction().id());
+        } else {
+          s = MissingTransactionStatus(__func__);
+          return MakeStatusOnlyResult<RowStream>(s.status());
```

Comment:
Observation: It still seems strange to me to fail an operation that actually succeeded (despite the missing transaction ID), but I guess I'm happy in the knowledge that it never happens.